### PR TITLE
Authorize backfill and scheduler requests + workflow modification & deletes

### PR DESCRIPTION
**Pull request by danielnorberg**
# Hey, I just made a Pull Request!

## Description
* Authorize requests that can be used to run a workflow (with parameters).
* For consistency, also authorize requests for injecting events, retrying and halting.
* Also authorizes requests to update or delete a workflow based on already configured service account.

## Motivation and Context
* Continuation of #630.
* Being able to trigger a workflow with parameters might allow arbitrary action, depending on the implementation of the workflow, and should thus also be authorized in the same way as workflow creation/updating requests.

## Checklist for PR author(s)

- [x] Changes are covered by unit test
- [x] All tests pass
- [ ] Code coverage check passes
- [ ] Error handling is tested
- [ ] Errors are handled at the appropriate layer
- [ ] Errors that cannot be handled where they occur are propagated
- [ ] (optional) Changes are covered by system test
- [ ] Relevant documentation updated
- [ ] This PR has NO breaking change to public API
- [ ] This PR has breaking change to public API and it is documented

## Checklist for PR reviewer(s)

- [ ] This PR has been incorporated in release note for the coming version
- [ ] Risky changes introduced by this PR have been all considered

### CODE REVIEW

File: styx-common/src/main/java/com/spotify/styx/api/Middlewares.java [hunk 1]
```diff
@@ -240,14 +241,23 @@ private static String internalServerErrorReason(String requestId, Throwable t) {
...
-    });
+    return ar -> rc -> ar
+        .apply(rc)
+        .apply(auth(rc, authenticator));
+  }
+
+  public static <E, R> Middleware<Authenticated<EntityResponseHandler<E, R>>, SyncHandler<Response<ByteString>>> authedEntity(
+      Middleware<EntityResponseHandler<E, R>, SyncHandler<Response<ByteString>>> entityMiddleware,
+      Middleware<Requested<Authenticated<Response<ByteString>>>, SyncHandler<Response<ByteString>>> authed) {
+    return ar -> rc -> authed.apply(r -> ac -> entityMiddleware.apply(ar.apply(ac)).invoke(rc)).invoke(rc);
```
[the remark this answers is not in the document]
Comment (NatashaL) [reply styx-common/src/main/java/com/spotify/styx/api/Middlewares.java]: I see. Both sound like somewhat of an undertaking that doesn't belong in this PR. So I guess we roll with this for a while

**Comment by danielnorberg** [comment]
As for adding authorization to resource operations, I agree that should be done, however I think that might need to wait for a future access control & authorization revamp.

The authorization we're currently implementing is actually for _workflow service account usage_, and we're piggy backing on this for authorizing operations that are directly coupled to a specific workflow, which is a bit of a hack. Unfortunately this cannot be easily extended to resources as they do not have an associated service account (and it is unclear what that would mean). Another quirk is that we cannot enforce authorization on workflows that do not have a service account configured (!).

A proper access control scheme would involve implementing access control policies for styx entities. I.e. a mechanism for assigning owner/read/write/etc permissions to workflows, resources, etc. Additionally service accounts should perhaps also be represented in styx as entities with associated access control policies, and we would just use the current SA usage authorization scheme when a user registers/creates an SA in styx.

**Comment by danielnorberg** [comment]
Not sure why codacy is unhappy. It claims that `ServiceAccountUsageAuthorizer.java` has an issue but doesn't show any 🤷‍♂️

File: styx-common/src/main/java/com/spotify/styx/api/Middlewares.java [hunk 2]
```diff
@@ -240,14 +241,23 @@ private static String internalServerErrorReason(String requestId, Throwable t) {
...
-  public static Middleware<Requested<Authenticated<Response<?>>>, AsyncHandler<Response<ByteString>>> authed(
+  public static <T> Middleware<Requested<Authenticated<Response<T>>>, SyncHandler<Response<T>>> authed(
       RequestAuthenticator authenticator) {
-    return ar -> jsonAsync().apply(requestContext -> {
-      final Response<?> payload = ar
-          .apply(requestContext)
-          .apply(auth(requestContext, authenticator));
-      return completedFuture(payload);
-    });
+    return ar -> rc -> ar
```
[the remark this answers is not in the document]
Comment (NatashaL) [reply styx-common/src/main/java/com/spotify/styx/api/Middlewares.java]: @danielnorberg can you comment on this?

[the remark this answers is not in the document]
Comment (danielnorberg) [reply styx-common/src/main/java/com/spotify/styx/api/Middlewares.java]: The change from `AsyncHandler` to `SyncHandler` was mostly to make it possible to use with the `EntityMiddleware` and compose with the `json()` middleware.

Apart from that it should work as before.

**Pull request merged**